main.go

```diff
@@ -300,45 +323,99 @@ func executeCLI(cmd *cobra.Command, src *source, w io.Writer) error {
 		return fmt.Errorf("unable to create renderer: %w", err)
 	}
 
-	content := string(b)
-	ext := filepath.Ext(src.URL)
-	if isCode {
-		content = utils.WrapCodeBlock(string(b), ext)
-	}
+	// Stream with incremental rendering using context
+	renderFunc := func(content []byte) []byte {
+		// Detect truly pathological patterns that will hang glamour
+		contentLen := len(content)
 
-	out, err := r.Render(content)
-	if err != nil {
-		return fmt.Errorf("unable to render markdown: %w", err)
+		// Check for excessive unclosed brackets that cause O(n²) parsing
+		if contentLen > 10000 {
+			// Count unclosed brackets - these cause exponential parsing time
+			openBrackets := bytes.Count(content, []byte("["))
+			closeBrackets := bytes.Count(content, []byte("]"))
+			unclosed := openBrackets - closeBrackets
+
+			// If more than 10% unclosed brackets in large content, skip rendering
+			if unclosed > contentLen/10 {
+				// Return the content as-is wrapped in a code block to prevent hang
+				return []byte(fmt.Sprintf("```\n%s\n```\n", content))
+			}
+		}
+
+		rendered, _ := r.RenderBytes(content)
+		return rendered
 	}
 
-	// display
+	// Determine output mode and handle accordingly
 	switch {
 	case pager || cmd.Flags().Changed("pager"):
+		// Buffer output for pager
+		var buf bytes.Buffer
+		err = Flow(ctx, src.reader, &buf, flow, renderFunc)
+		if err != nil {
+			return err
+		}
+
+		// Send buffered output to pager
 		pagerCmd := os.Getenv("PAGER")
 		if pagerCmd == "" {
 			pagerCmd = "less -r"
 		}
-
 		pa := strings.Split(pagerCmd, " ")
 		c := exec.Command(pa[0], pa[1:]...) //nolint:gosec
-		c.Stdin = strings.NewReader(out)
+		c.Stdin = bytes.NewReader(buf.Bytes())
 		c.Stdout = os.Stdout
+		c.Stderr = os.Stderr
 		if err := c.Run(); err != nil {
-			return fmt.Errorf("unable to run command: %w", err)
+			return fmt.Errorf("unable to run pager: %w", err)
 		}
 		return nil
+
 	case tui || cmd.Flags().Changed("tui"):
+		// Read all content for TUI
+		b, err := io.ReadAll(src.reader)
+		if err != nil {
+			return fmt.Errorf("unable to read content: %w", err)
+		}
+
+		b = utils.RemoveFrontmatter(b)
+		content := string(b)
+		ext := filepath.Ext(src.URL)
+		if isCode {
+			content = utils.WrapCodeBlock(string(b), ext)
+		}
+
+		// Render content for TUI
+		renderedContent, err := r.Render(content)
+		if err != nil {
+			return fmt.Errorf("unable to render content: %w", err)
+		}
+
 		path := ""
 		if !isURL(src.URL) {
 			path = src.URL
 		}
-		return runTUI(path, content)
+		return runTUI(path, renderedContent)
+
 	default:
-		if _, err = fmt.Fprint(w, out); err != nil {
-			return fmt.Errorf("unable to write to writer: %w", err)
+		// Stream directly to output (preserves streaming performance)
+		// Wrap stdout in a bufio.Writer so we can flush it
+		bw := bufio.NewWriter(w)
+		defer bw.Flush()
+		err = Flow(ctx, src.reader, bw, flow, renderFunc)
+	}
+
+	// If we received a signal, exit with appropriate signal exit code
```

Comment:
This definitely needs to be moved to the top of the main function, or it will skip `defers` lower in the stack.